docs/adr/0007-preventing-version-mismatches.md

```diff
@@ -0,0 +1,269 @@
+    change. Our custom/external types work would not.
+  - For code generators: the minor version is incremented whenever there's
+    a breaking change between the generated code and the consumer code.  Some
+    examples would be: changing Kotlin classes from open to closed and changing
+    the argument names on Swift.
+  - For `uniffi` itself and other crates, I'm not sure what a minor version
+    change should mean.  Maybe we just don't bump these?
+  - The patch version is incremented for non-breaking, bugfix, releases.
+
+This means that `uniffi-bindgen` and all external bindgen tools are compatible
+with `uniffi` if their major versions match.  We would update
+`ComponentInterface` so it only uses the major version number when calculating
+it's hash.
+
+Pros:
+  - Seems like a hard requirement if we want external bindings tools
+  - Would significantly reduce the chance of running into a version mismatch.
+  - Makes it easier to detect version mismatches when parsing `Cargo.toml`.
+    Right now we need to determine the exact version that was resolved and
+    stored in `Cargo.lock`.  If we only needed to compare the major versions,
+    then we could use the specification from `Cargo.toml` directly.
+
+Cons
+  - We would have to be more careful with breaking changes. If we realized a
+    change would break the generated code compatibility, then that change might
+    need to wait before releasing it. If we didn't realize a change would break
+    the generated code compatilibity and released the change anyway, then we've
+    introduced undefined behavior in all consumer code.
+  - Doesn't fully solve the issue of version mismatches. We would need to
+    combine this with at one or more options below.
+
+### **[Option 1] generate bindings in `build.rs`**
+
+As mentioned in the context, we handle this issue for scaffolding by using the
+`builtin-bindgen` feature of `uniffi_build`.  If we also generated the bindings
+code in `build.rs`, we could use a similar system:
+
+- The Rust crate adds a dependency to each bindings generator they want to
+  use (`uniffi-bindgen-kotlin = 1.10.1`, `uniffi-bindgen-swift = 1.4.0`,
+  etc).
+- Each bindings generator defines a `generate_bindings()` function, which
+  generates the bindings and writes them to the `target/[language]` directory.
+- We call this in `build.rs`:
+
+```
+fn main() {
+  uniffi_bindgen_language1.generate_bindings();
+  uniffi_bindgen_language2.generate_bindings();
+  ...
+}
+```
+
+- The build scripts would then copy the bindings from that directory to the
+  final location.
+
+Pros:
+  - Doesn't rely on a system-wide install of `uniffi-bindgen`.
+  - Bindgen tool versions are specified in the `Cargo.toml`, alongside the
+    `uniffi` version.
+
+Cons:
+  - Forces a specific order to the build process.  If you try building the
+    bindings before the Rust component, it would fail since the source files
+    haven't been generated yet (or worse, succeed with stale code).
```

Comment:
it's worth thinking about this in a future where .udl files didn't exist and macros are used - that world might make this less of a problem and possibly even a feature? eg, assuming foreign bindings need to parse rust code, it might actually be a build time benefit to generating all bindings at the same time from the same parse tree?

[ok, I've now moved from "thinking out loud" to "making things up" :]